drivers: comparator: Add Comparator support on Renesas RX

quytranpzz opened the pull request:
Add comparator driver support on Renesas RX130 using LVD peripheral.
This driver uses a voltage threshold as the negative input, while the positive input can be selected between Vcc or the CMPA pin.

thenguyenyf commented on dts/bindings/misc/renesas,rx-lvd.yaml:
```diff
@@ -0,0 +1,82 @@
+    type: int
+    description: RX LVD channel
+
+  lvd-action:
+    required: true
+    type: int
+    enum: [0, 1, 2, 3]
+    description: |
+      Choose the action to be taken when the LVD is detected.
+        - 0: Reset
+        - 1: Non-maskable interrupt
+        - 2: Maskable interrupt
+        - 3: No action
```
Maybe better for readability
```suggestion
  lvd-action:
    required: true
    type: int
    enum:
      - "reset"
      - "non-maskable-interrupt"
      - "maskable-interrupt"
      - "no-action"
    description: |
      Choose the action to be taken when the LVD is detected.
```

quytranpzz replied:
It's done

thenguyenyf commented on drivers/misc/renesas_rx_lvd/Kconfig:
```diff
@@ -0,0 +1,19 @@
+	help
+	  Enable Renesas RX low voltage detection driver
+
+config RENESAS_RX_LVD_INIT_PRIORITY
+	int "RX LVD initialization priority"
+	default KERNEL_INIT_PRIORITY_DEVICE
+	help
+	  System initialization priority for RX LVD driver.
```
Should be wrapped by `RENESAS_RX_LVD`
```suggestion
if RENESAS_RX_LVD

config RENESAS_RX_LVD_INIT_PRIORITY
	int "RX LVD initialization priority"
	default KERNEL_INIT_PRIORITY_DEVICE
	help
	  System initialization priority for RX LVD driver.

endif # RENESAS_RX_LVD
```

thenguyenyf commented on drivers/misc/renesas_rx_lvd/renesas_rx_lvd.c:
```diff
@@ -0,0 +1,206 @@
+		.callback = rx_lvd_callback_##index,                                               \
+	};                                                                                         \
+                                                                                                   \
+	static int renesas_rx_lvd_init_##index(const struct device *dev)                           \
+	{                                                                                          \
+		int err = renesas_rx_lvd_init(dev);                                                \
+		if (err != 0) {                                                                    \
+			return err;                                                                \
+		}                                                                                  \
+		return 0;                                                                          \
+	}                                                                                          \
```
It seems this wrapper is redundant. Why not use `renesas_rx_lvd_init` directly in `DEVICE_DT_INST_DEFINE`?

quytranpzz replied:
I removed this. thanks

thenguyenyf commented on dts/bindings/misc/renesas,rx-lvd.yaml:
```diff
@@ -0,0 +1,82 @@
+        - 2: Maskable interrupt
+        - 3: No action
+
+  voltage-level:
+    required: true
+    type: int
+    description: |
+      Specifies the voltage detection level for each channel, set an integer
+      value which expresses the number up to two decimal places.
+      Example:
+        - To set the voltage detection level to 3.00 V, specify '300'.
+        - To set the voltage detection level to default value, specify '0xFFFF'
+      For specific voltage detection support levels of each RX MCU series,
+      please refer to the HWM.
```
I'm not a fan of this kind of setting. I think an array to separate integer and decimal parts, something like https://github.com/zephyrproject-rtos/zephyr/blob/2aeac6087d8a05726c8e3cacc22f0ce4f0382490/boards/renesas/ek_ra8m1/ek_ra8m1.dts#L164 would be better

thenguyenyf commented on include/zephyr/drivers/misc/renesas_rx_lvd/renesas_rx_lvd.h:
```diff
@@ -0,0 +1,67 @@
+ * @retval 0 On success.
+ * @retval -EINVAL On failure.
+ */
+int renesas_rx_pin_set_cmpa2(const struct device *dev);
```
Is it necessary for an application call?

quytranpzz commented on include/zephyr/drivers/misc/renesas_rx_lvd/renesas_rx_lvd.h:
I added a new header file for Renesas LVD driver support. Please take a look

thenguyenyf reviewed (dismissed):
LGTM

kartben reviewed (dismissed):
-1 until the comparator discussion happens. I will also add comments re: a few things that need fixing in the docs and sample. Thanks!

bjarki-andreasen commented on include/zephyr/drivers/misc/renesas_lvd/renesas_lvd.h:
```diff
@@ -0,0 +1,155 @@
+ *  This enum indicates whether the monitored voltage is
+ *  currently above or below the configured threshold.
+ */
+typedef enum {
+	/** Voltage is above threshold. */
+	RENESAS_LVD_POSITION_ABOVE = 0,
+
+	/** Voltage is below threshold. */
+	RENESAS_LVD_POSITION_BELOW,
+} renesas_lvd_position_t;
```
This maps well to https://github.com/zephyrproject-rtos/zephyr/blob/90723c86ed7f4166bb94cbf2c65f1e127e51977d/include/zephyr/drivers/comparator.h#L60-L74

bjarki-andreasen commented on include/zephyr/drivers/misc/renesas_lvd/renesas_lvd.h:
```diff
@@ -0,0 +1,155 @@
+ *  This enum defines the status, whether the voltage crossed
+ *  the voltage detection level or not
+ */
+typedef enum {
+	/** No threshold crossing detected */
+	RENESAS_LVD_CROSS_NONE = 0,
+
+	/** Voltage crossed over the threshold. */
+	RENESAS_LVD_CROSS_OVER,
+} renesas_lvd_cross_t;
```
This maps to https://github.com/zephyrproject-rtos/zephyr/blob/90723c86ed7f4166bb94cbf2c65f1e127e51977d/include/zephyr/drivers/comparator.h#L27-L36 specifically https://github.com/zephyrproject-rtos/zephyr/blob/90723c86ed7f4166bb94cbf2c65f1e127e51977d/include/zephyr/drivers/comparator.h#L31 which seems to be the only edge this hardware supports

bjarki-andreasen commented on include/zephyr/drivers/misc/renesas_lvd/renesas_lvd.h:
```diff
@@ -0,0 +1,155 @@
+ * (Internal use only.)
+ */
+__subsystem struct renesas_lvd_driver_api {
+	int (*get_status)(const struct device *dev, renesas_lvd_status_t *status);
+	int (*clear_status)(const struct device *dev);
+	int (*callback_set)(const struct device *dev, renesas_lvd_callback_t callback,
+			    void *user_data);
```
These map to https://github.com/zephyrproject-rtos/zephyr/blob/90723c86ed7f4166bb94cbf2c65f1e127e51977d/include/zephyr/drivers/comparator.h#L51-L56
`get_status` -> `get_output`, `callback_set` -> `set_trigger_callback`, `clear_status` -> `trigger_is_pending`. The comparator API additionally has the `set_trigger` API which allows the user to enable/disable the edge detection (interrupt).

bjarki-andreasen replied:
A device does not need to meet the full functionality of a device driver API, most devices in fact don't :) Drivers return `-ENOTSUP` if the application tries to make a specific device do something which it does not support, so unless the LVD has features which are not covered by the comparator API, which it seems it does not, its perfectly fit to use the comparator API to implement the features it does support

quytranpzz replied:
On top of that, the current LVD usage description does not match the comparator description and may confuse users.
- Comparator description: "compares the voltages of two analog signals connected to its negative and positive inputs."
- LVD description: "Compare VCC supply with voltage threshold."

quytranpzz replied:
Yes, but the LVD usage does not match the comparator description and may confuse users.
comparator: "Compares the voltages of two analog signals connected to its negative and positive inputs"
LVD just "compares VCC supply with voltage threshold"

bjarki-andreasen replied:
Right, the voltage threshold is the negative input, and VCC the positive input, in which case if VCC is above the threshold, the output will be 1 (high) :) Just because the inputs can't be freely selected it does not make the hardware "not a comparator" does it?

bjarki-andreasen reviewed (changes requested):
This hardware maps really well to the comparator subsystem, check out https://docs.zephyrproject.org/latest/hardware/peripherals/comparator.html , and see the comments I added to this PR. The comparator subsystem comes with documentation, a shell, and test suites, so you get a lot "for free" including portability using it instead of introducing a new API :)

There is a component in tree which contains multiple comparator's as well, check out https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/comparator/renesas%2Cra-acmphs-global.yaml and https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/comparator/renesas%2Cra-acmphs.yaml Comparators don't have "channels", a comparator is "a comparator" if there are two channels, there are two comparators :)

quytranpzz committed:
drivers: comparator: Add comparator driver support for RX

Add comparator support for Renesas RX with LVD

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

quytranpzz committed:
dts: renesas: rx: Add dts property nodes for LVD support

Add DTS node for LVD support on RX130

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

quytranpzz committed:
samples: boards: renesas: Add LVD sample for Renesas

Add a simple sample using Comparator API to monitor the
voltage on Renesas boards

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

quytranpzz committed:
manifest: hal: renesas: Update commit ID for Renesas HAL

Update the latest commit ID for HAL Renesas

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

decsny commented:
this is literally the 3rd PR I looked at in 15 minutes where NXP maintainer got pulled in despite no NXP files, please fix the MAINTAINER.yml

quytranpzz commented on drivers/comparator/CMakeLists.txt:
```diff
@@ -14,3 +14,4 @@ zephyr_library_sources_ifdef(CONFIG_COMPARATOR_NRF_COMP comparator_nrf_comp.c)
 zephyr_library_sources_ifdef(CONFIG_COMPARATOR_NRF_LPCOMP comparator_nrf_lpcomp.c)
 zephyr_library_sources_ifdef(CONFIG_COMPARATOR_SHELL comparator_shell.c)
 zephyr_library_sources_ifdef(CONFIG_COMPARATOR_RENESAS_RA comparator_renesas_ra.c)
+zephyr_library_sources_ifdef(CONFIG_COMPARATOR_RENESAS_RX_LVD comparator_renesas_rx.c)
```
It's done
